File: packages/sdk/src/emitter.ts
import type { EdgeFlagsEvent, EventPayloadMap } from './types.js';

type Listener<E extends EdgeFlagsEvent> = (payload: EventPayloadMap[E]) => void;

export class Emitter {
  private listeners = new Map<EdgeFlagsEvent, Set<Listener<never>>>();

  on<E extends EdgeFlagsEvent>(event: E, fn: Listener<E>): () => void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(fn as Listener<never>);

    return () => {
      this.listeners.get(event)?.delete(fn as Listener<never>);
    };
  }

  emit<E extends EdgeFlagsEvent>(event: E, payload: EventPayloadMap[E]): void {
    const fns = this.listeners.get(event);
    if (!fns) return;
    for (const fn of fns) {
      (fn as Listener<E>)(payload);
    }
  }

  removeAll(): void {
    this.listeners.clear();
  }
}

File: packages/python/src/edgeflags/emitter.py
from __future__ import annotations

import threading
from collections.abc import Callable
from typing import Any


class Emitter:
    def __init__(self) -> None:
        self._listeners: dict[str, list[Callable[..., Any]]] = {}
        self._lock = threading.Lock()

    def on(self, event: str, fn: Callable[..., Any]) -> Callable[[], None]:
        with self._lock:
            if event not in self._listeners:
                self._listeners[event] = []
            self._listeners[event].append(fn)

        def unsubscribe() -> None:
            with self._lock:
                listeners = self._listeners.get(event)
                if listeners and fn in listeners:
                    listeners.remove(fn)

        return unsubscribe

    def emit(self, event: str, payload: Any = None) -> None:
        with self._lock:
            listeners = list(self._listeners.get(event, []))
        for fn in listeners:
            if payload is None:
                fn()
            else:
                fn(payload)

    def remove_all(self) -> None:
        with self._lock:
            self._listeners.clear()
